binaries/bzutil/main.go

```diff
@@ -342,3 +388,101 @@ func printSupported() {
 	}
 	fmt.Printf("Supported commands: %s\n", strings.Join(cmds, ", "))
 }
+
+// use batch upload, print the list of digests that were uploaded
+func batchUploadFiles(casAddr string, uploadDir string, testUpload bool) (string, error) {
+	casClient := cas.MakeCASClient()
+	resolver := dialer.NewConstantResolver(casAddr)
+	// set up the contents for upload
+	contents, e := getContents(uploadDir, testUpload)
+	if e != nil {
+		return "", e
+	}
+
+	// request the upload
+	digests, e := casClient.BatchUpdateWrite(resolver, contents, &backoff.ConstantBackOff{Interval:time.Millisecond})
+	if e != nil {
+		return "", e
+	}
+
+	// handle response
+	asStr, e := json.Marshal(digests)
+	return string(asStr), e
+}
+
+// get batch upload contents from either the files in a directory, or test set of 10 blobs
+func getContents(dir string, test bool) ([]cas.BatchUploadContent, error) {
+	contents := make([]cas.BatchUploadContent, 0)
+	if !test {
+		fileList, e := ioutil.ReadDir(dir)
+		if e != nil {
+			return nil, e
+		}
+		for _, f := range fileList {
+			if ! f.IsDir() {
+				data, e := ioutil.ReadFile(f.Name())
+				if e != nil {
+					return nil, fmt.Errorf("error reading %s:%s", f.Name(), e.Error())
+				}
+				uploadContent := makeUploadContent(data)
+				contents = append(contents, uploadContent)
+			}
+		}
+		return contents, nil
+	}
+
+	for i := 0; i < 10; i++ {
+		data := make([]byte, (i+1)*1000)
+		rand.Read(data)
+		uploadContent := makeUploadContent(data)
+		contents = append(contents, uploadContent)
+	}
+	return contents, nil
+}
+
+func makeUploadContent(data []byte) (cas.BatchUploadContent) {
```

Comment:
This feels like we are conflating the perftest tool with bzutil. bzutil should ideally be as thin a wrapper around remexec client actions as possible. In that case bzutil would just implement the core functionality, and any wrapper tool would be where inputs were generated. The read subcommand is close to what we want although it does deviate by using json inputs. Everything else expects a digest to be in the form "<hash>/<size>".

Reply:
this doesn't make the data being uploaded... its making the BatchUploadContent data structure used to upload a batch of data blobs.  (I'll rename) BatchUploadContent is custom to bzutil (our bazel client utility to allow bazel users to provide a single list of (digests, data blob) pairs.  
If we don't want to introduce a new data structure we can simply use 2 lists (a list of digests and a list of data blobs).  I consider 2 lists to be more fragile: the list producers need to ensure that the 2 sets of entries are aligned (make sure they don't have a bug that adds one entry to the list but not to the other).

Reply:
Ok, sorry, I think I misread what this is doing. Seems reasonable as is